sim/midas/src/main/scala/midas/widgets/Widget.scala

```diff
@@ -31,27 +32,37 @@ object WidgetMMIO {
 class WidgetIO(implicit p: Parameters) extends ParameterizedBundle()(p){
   val ctrl = Flipped(WidgetMMIO())
 }
+abstract class Widget(implicit p: Parameters) extends LazyModule {
+  override def module: WidgetImp
+  val wName = Some(Widget.assignName(this))
+  // Legacy API
+  def getWName = wName.get
 
-abstract class Widget(implicit val p: Parameters) extends MultiIOModule {
-  private var _finalized = false
-  protected val crRegistry = new MCRFileMap()
-  def numRegs = crRegistry.numRegs
+  // Returns widget-relative word address
+  def getCRAddr(name: String): Int = {
+    module.crRegistry.lookupAddress(name).getOrElse(
+      throw new RuntimeException(s"Could not find CR:${name} in widget: $wName"))
+  }
 
-  def io: WidgetIO
+  def headerComment(sb: StringBuilder) {
+    val name = getWName.toUpperCase
+    sb.append("\n// Widget: %s\n".format(getWName))
+    sb.append(CppGenerationUtils.genMacro(s"${name}(x)", s"${name}_ ## x"))
+  }
 
   val customSize: Option[BigInt] = None
-  // Default case we set the region to be large enough to hold the CRs
-  lazy val memRegionSize = customSize.getOrElse(
-    BigInt(1 << log2Up(numRegs * (io.ctrl.nastiXDataBits/8))))
+  def memRegionSize = customSize.getOrElse(BigInt(1 << log2Up(module.numRegs * (module.io.ctrl.nastiXDataBits/8))))
 
-  protected var wName: Option[String] = None
-  private def setWidgetName(n: String) {
-    wName = Some(n)
-  }
-  def getWName: String = {
-    wName.getOrElse(throw new  RuntimeException("Must build widgets with their companion object"))
-  }
+  def printCRs = module.crRegistry.printCRs
+}
+
+abstract class WidgetImp(wrapper: Widget)(implicit p: Parameters) extends LazyModuleImp(wrapper) {
```

Comment:
Is the implicit parameters argument necessary here? Usually module implementations don't declare their own implicit parameters object, since they will use the one from the parent lazy module.

Reply:
Good point, will do.